libraries/rush-lib/src/logic/operations/BuildPlanPlugin.ts

```diff
@@ -0,0 +1,298 @@
+      >();
+      for (const operation of operations) {
+        const { associatedProject, associatedPhase } = operation;
+        if (associatedProject && associatedPhase) {
+          const projectConfiguration: RushProjectConfiguration | undefined =
+            projectConfigurations.get(associatedProject);
+          const fileHashes: Map<string, string> | undefined =
+            await projectChangeAnalyzer._tryGetProjectDependenciesAsync(associatedProject, terminal);
+          const cacheDisabledReason: string | undefined = projectConfiguration
+            ? projectConfiguration.getCacheDisabledReason(fileHashes!.keys(), associatedPhase.name, operation)
+            : `Project does not have a ${RushConstants.rushProjectConfigFilename} configuration file, ` +
+              'or one provided by a rig, so it does not support caching.';
+          buildCacheByOperation.set(operation, { cacheDisabledReason });
+        }
+      }
+      clusterOperations(disjointSet, buildCacheByOperation);
+      const buildPlan: ICobuildPlan = createCobuildPlan(disjointSet, terminal, buildCacheByOperation);
+      logCobuildBuildPlan(buildPlan, terminal);
+    }
+  }
+}
+
+function generateCobuildPlanSummary(operations: Operation[], terminal: ITerminal): ICobuildPlan['summary'] {
+  const noOpOperations: Set<Operation> = new Set(operations.filter((e) => e.runner?.isNoOp));
+  const leafQueue: Operation[] = [];
+
+  for (const operation of operations) {
+    if (noOpOperations.has(operation)) {
+      continue;
+    }
+    let numberOfConsumers: number = 0;
+    for (const consumer of operation.consumers) {
+      if (!noOpOperations.has(consumer)) {
+        numberOfConsumers++;
+      }
+    }
+    if (numberOfConsumers === 0) {
+      leafQueue.push(operation);
+    }
+  }
+  let currentLeafNodes: Set<Operation> = new Set<Operation>();
+  const remainingOperations: Set<Operation> = new Set<Operation>(operations);
+  let depth: number = 0;
+  let maxWidth: number = leafQueue.filter((e) => !e.runner?.isNoOp).length;
+  const numberOfNodes: number[] = [maxWidth];
+  const depthToOperationsMap: Map<number, Set<Operation>> = new Map<number, Set<Operation>>();
+  depthToOperationsMap.set(depth, new Set(leafQueue));
+  do {
+    if (leafQueue.length === 0) {
+      leafQueue.push(...currentLeafNodes);
+      const realOperations: Operation[] = leafQueue.filter((e) => !e.runner?.isNoOp);
+      if (realOperations.length > 0) {
+        depth += 1;
+        depthToOperationsMap.set(depth, new Set(realOperations));
+        numberOfNodes.push(realOperations.length);
+      }
+      const currentWidth: number = realOperations.length;
+      if (currentWidth > maxWidth) {
+        maxWidth = currentWidth;
+      }
+      currentLeafNodes = new Set();
+    }
+    if (leafQueue.length === 0) {
+      break;
+    }
+    const leaf: Operation = leafQueue.shift()!;
+    if (remainingOperations.has(leaf)) {
+      remainingOperations.delete(leaf);
```

Comment:
```suggestion
    if (remainingOperations.delete(leaf)) {
```
`delete` returns a `boolean` indicating if it did anything.